src/java/za/co/mip/ablduck/ABLDuck.java

```diff
@@ -0,0 +1,345 @@
+import com.openedge.pdt.core.ast.model.IASTContext;
+import com.openedge.pdt.core.ast.model.ICompilationUnit;
+
+import com.phenix.pct.PCT;
+import com.phenix.pct.Messages;
+import com.phenix.pct.Version;
+
+import za.co.mip.ablduck.models.DataJSObject;
+import za.co.mip.ablduck.models.SourceJSObject;
+import za.co.mip.ablduck.models.data.ClassDataObject;
+import za.co.mip.ablduck.models.data.SearchDataObject;
+import za.co.mip.ablduck.models.source.MemberObject;
+import za.co.mip.ablduck.utilities.HTMLGenerator;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Class for generating ABLDuck documentation from OpenEdge classes
+ * 
+ * @author <a href="mailto:robertedwardsmail@gmail.com">Robert Edwards</a>
+ */
+public class ABLDuck extends PCT {
+    private HashMap<String, SourceJSObject> jsObjects = new HashMap<>();
+    private String title = "ABLDuck documentation";
+    private File destDir = null;
+    private File destDirOutput = null;
+    private List<FileSet> filesets = new ArrayList<>();
+    protected Path propath = null;
+
+    public ABLDuck() {
+        super();
+        createPropath();
+    }
+
+	/**
+     * Adds a set of files to archive.
+     * 
+     * @param set FileSet
+     */
+    public void addFileset(FileSet set) {
+        filesets.add(set);
+    }
+
+    /**
+     * Destination directory
+     * 
+     * @param destFile Directory
+     */
+    public void setDestDir(File dir) {
+        this.destDir = dir;
+        this.destDirOutput = new File(dir, "output");
+    }
+
+    /**
+     * Documentation title
+     * 
+     * @param title Documentation title
+     */
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    /**
+     * Set the propath to be used when running the procedure
+     * 
+     * @param propath an Ant Path object containing the propath
+     */
+    public void addPropath(Path propath) {
+        createPropath().append(propath);
+    }
+
+    /**
+     * Creates a new Path instance
+     * 
+     * @return Path
+     */
+    public Path createPropath() {
+        if (this.propath == null) {
+            this.propath = new Path(this.getProject());
+        }
+
+        return this.propath;
+    }
+
+    @Override
+    public void execute() {
+        checkDlcHome();
+
+        // Destination directory must exist
+        if (this.destDir == null) {
+            throw new BuildException(MessageFormat.format(Messages.getString("OpenEdgeClassDocumentation.0"), "destDir"));
+        }
+        // There must be at least one fileset
+        if (filesets.isEmpty()) {
+            throw new BuildException(Messages.getString("OpenEdgeClassDocumentation.1"));
+        }
+
+        this.destDirOutput.mkdirs();
+
+        //Extract template
+        try {
+            extractTemplateDirectory(this.destDir);
+
+            replaceTemplateTags("{title}", this.title, Paths.get(this.destDir.getAbsolutePath(), "index.html"));

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "{title}" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1192)
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "index.html" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1192)